vault/identity_store_structs.go

```diff
@@ -0,0 +1,184 @@
+package vault
+
+import (
+	"regexp"
+	"time"
+
+	memdb "github.com/hashicorp/go-memdb"
+	"github.com/hashicorp/vault/helper/locksutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	log "github.com/mgutz/logxi/v1"
+)
+
+const (
+	// Storage prefixes
+	entityPrefix = "entity/"
+)
+
+var (
+	// metaKeyFormatRegEx checks if a metadata key string is valid
+	metaKeyFormatRegEx = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`).MatchString
+)
+
+const (
+	// The meta key prefix reserved for Vault's internal use
+	metaKeyReservedPrefix = "vault-"
```

Comment:
Good idea. In fact, by using this we don't really need to consider any difference between implicit/explicit metadata. We can just know that only Vault can set metadata with `vault-` and that *all* Vault-generated metadata will start with it.